### PR TITLE
Event deletion

### LINKED ISSUE #240: fix: deleting an event gives error
How to produce the error:
1. Create an Event
2. Create a sponsor for that event
3. Now try to delete that event.It will give an Internal Server Error.
This will happen when an event has either a sponsor, session, microlocation ,tracks etc...
It will work successfully if only an event is created. Basically this happens because of the ForeignKey constraint on the sponsor,session ,microlocation,etc... objects which prevent the event from deleting.

### PR DESCRIPTION
**Pull request by SaptakS**
Fixes #240

### CODE REVIEW

**Comment by SaptakS** [comment]
@rafalkowalski please review

**Comment by aditya1702** [comment]
@SaptakS this isnt working.I am still getting the same error

**Comment by SaptakS** [comment]
It was working fine for me.. Can you give a screenshot of the logs in console?

**Comment by aditya1702** [comment]
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python2.7/dist-packages/flask_cors/extension.py", line 188, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/local/lib/python2.7/dist-packages/flask_admin/base.py", line 68, in inner
    return self._run_view(f, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/flask_admin/base.py", line 354, in _run_view
    return fn(self, *args, **kwargs)
  File "/vagrant/open_event/views/admin/models_views/event.py", line 131, in delete_view
    DataManager.delete_event(event_id)
  File "/vagrant/open_event/helpers/data.py", line 464, in delete_event
    Event.query.filter_by(id=e_id).delete()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/query.py", line 3048, in delete
    delete_op.exec_()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/persistence.py", line 1127, in exec_
    self._do_exec()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/persistence.py", line 1311, in _do_exec
    mapper=self.mapper)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 1034, in execute
    bind, close_with_result=True).execute(clause, params or {})
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 914, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/sql/elements.py", line 323, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1010, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1146, in _execute_context
    context)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1341, in _handle_dbapi_exception
    exc_info
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/util/compat.py", line 200, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1139, in _execute_context
    context)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/default.py", line 450, in do_execute
    cursor.execute(statement, parameters)
IntegrityError: (psycopg2.IntegrityError) update or delete on table "events" violates foreign key constraint "sponsors_event_id_fkey" on table "sponsors"
DETAIL:  Key (id)=(3) is still referenced from table "sponsors".
 [SQL: 'DELETE FROM events WHERE events.id = %(id_1)s'] [parameters: {'id_1': u'3'}]
10.0.2.2 - - [10/Apr/2016 18:58:42] "POST /admin/event/delete/ HTTP/1.1" 500 -

**Comment by SaptakS** [comment]
According to this, the entry from sponsor table isnt deleted but I have deleted the entry from sponsor table. Anyways.. I will have a look..

**Comment by aviaryan** [comment]
@aditya1702 I think the reason for your error is that @SaptakS 's changes has not been pulled correctly to your local repo. I can say this for sure because your trace has
```
File "/vagrant/open_event/helpers/data.py", line 464, in delete_event
Event.query.filter_by(id=e_id).delete()
```
In Saptak's commits, that line has moved to 508 ([diff link](https://github.com/fossasia/open-event-orga-server/pull/241/commits/4b92dabefb8ca3b7ecb3cf24ab419fd68c737f95#diff-9849360ece7f55a816d9268d417ce9e5R508)).

File: open_event/helpers/data.py [hunk 1]
```diff
@@ -71,11 +72,48 @@ def remove_track(track_id):
...
+        """
+        review will be saved to database with proper Event id
+        :param data: review data
+        :param session_id: review belongs to Event by event id
+        """
+        new_review = Review(email=data["email"],
+                              rating=data["rating"],
+                              comment=data["comment"],
+                              session_id=session_id)
+
```
Comment (juslee) [inline open_event/helpers/data.py]: Codacy Issue found: [Trailing whitespace](https://www.codacy.com/app/triplez/open-event-orga-server/file/2406546284/issues/source?bid=3233864&fileBranchId=3233880#l91)

File: open_event/helpers/data.py [hunk 2]
```diff
@@ -71,11 +72,48 @@ def remove_track(track_id):
...
+                              comment=data["comment"],
+                              session_id=session_id)
+        
+        save_to_db(new_review, "review saved")
+
+    @staticmethod
+    def update_review(data, review, session_id):
+        """
+        review will be updated in database
+        :param data: review data
```
Comment (juslee) [inline open_event/helpers/data.py]: Codacy Issue found: [Trailing whitespace](https://www.codacy.com/app/triplez/open-event-orga-server/file/2406546284/issues/source?bid=3233864&fileBranchId=3233880#l98)

File: open_event/views/views.py [hunk 3]
```diff
@@ -278,6 +282,44 @@ def generate_icalender_track(track_id):
...
+                 'detail': 'Review Updated'})
+
+    elif "email" in data and "rating" in data: 
+        if not re.match(r"[^@]+@[^@]+\.[^@]+", data["email"]):
+            return jsonify(
+                {'status': '400',
+                 'detail': 'Not valid data'})
+        DataManager.create_review(data, session_id)        
+        return jsonify(
+                {'status': '200',
```
Comment (juslee) [inline open_event/views/views.py]: Codacy Issue found: [Trailing whitespace](https://www.codacy.com/app/triplez/open-event-orga-server/file/2406546320/issues/source?bid=3233864&fileBranchId=3233880#l317)

File: open_event/views/views.py [hunk 4]
```diff
@@ -278,6 +282,44 @@ def generate_icalender_track(track_id):
...
+        DataManager.update_review(data, reviews[0], session_id)
+        return jsonify(
+                {'status': '200',
+                 'detail': 'Review Updated'})
+
+    elif "email" in data and "rating" in data: 
+        if not re.match(r"[^@]+@[^@]+\.[^@]+", data["email"]):
+            return jsonify(
+                {'status': '400',
+                 'detail': 'Not valid data'})
```
Comment (juslee) [inline open_event/views/views.py]: Codacy Issue found: [Trailing whitespace](https://www.codacy.com/app/triplez/open-event-orga-server/file/2406546320/issues/source?bid=3233864&fileBranchId=3233880#l314)

File: open_event/views/views.py [hunk 5]
```diff
@@ -278,6 +282,44 @@ def generate_icalender_track(track_id):
...
+    if "comment" not in data:
+        data["comment"] = ""
+    reviews = DataGetter.get_reviews_by_email_and_session_id(data["email"])
+
+    if reviews.count() > 0:
+        DataManager.update_review(data, reviews[0], session_id)
+        return jsonify(
+                {'status': '200',
+                 'detail': 'Review Updated'})
+
```
Comment (juslee) [inline open_event/views/views.py]: Codacy Issue found: [Trailing whitespace](https://www.codacy.com/app/triplez/open-event-orga-server/file/2406546320/issues/source?bid=3233864&fileBranchId=3233880#l309)

File: open_event/views/views.py [hunk 6]
```diff
@@ -9,8 +10,11 @@
 from ..models.microlocation import Microlocation
 from ..models.event import Event
 from ..models.session import Session, Level, Format, Language
+from ..models.reviews import Review
 from ..models.version import Version
 from ..helpers.object_formatter import ObjectFormatter
+from ..helpers.data import DataManager, save_to_db, delete_from_db
```
Comment (juslee) [inline open_event/views/views.py]: Codacy Issue found: [Unused save_to_db imported from helpers.data](https://www.codacy.com/app/triplez/open-event-orga-server/file/2406546320/issues/source?bid=3233864&fileBranchId=3233880#l16)

File: open_event/models/reviews.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+class Review(db.Model):
+    """Review model class"""
+    __tablename__ = 'session_reviews'
+    id = db.Column(db.Integer, primary_key=True)
+    email = db.Column(db.String, nullable=False)
+    comment = db.Column(db.String)
+    rating = db.Column(db.Integer, nullable=False)
+    session_id = db.Column(db.Integer, db.ForeignKey('session.id'))
+
```
Comment (juslee) [inline open_event/models/reviews.py]: Codacy Issue found: [Trailing whitespace](https://www.codacy.com/app/triplez/open-event-orga-server/file/2406546298/issues/source?bid=3233864&fileBranchId=3233880#l12)

File: open_event/models/reviews.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+from . import db
+from .session import Session
```
Comment (juslee) [inline open_event/models/reviews.py]: Codacy Issue found: [Unused Session imported from session](https://www.codacy.com/app/triplez/open-event-orga-server/file/2406546298/issues/source?bid=3233864&fileBranchId=3233880#l2)

File: migrations/versions/0b5368c41418_.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+"""
+
+# revision identifiers, used by Alembic.
+revision = '0b5368c41418'
+down_revision = '202b15987640'
+
+from alembic import op
+import sqlalchemy as sa
+import sqlalchemy_utils
```
Comment (juslee) [inline migrations/versions/0b5368c41418_.py]: Codacy Issue found: [Unused import sqlalchemy_utils](https://www.codacy.com/app/triplez/open-event-orga-server/file/2406546261/issues/source?bid=3233864&fileBranchId=3233880#l15)

File: migrations/versions/0b5368c41418_.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def upgrade():
+    ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('session_reviews',
+    sa.Column('id', sa.Integer(), nullable=False),
+    sa.Column('email', sa.String(), nullable=False),
```
Comment (juslee) [inline migrations/versions/0b5368c41418_.py]: Codacy Issue found: [Trailing whitespace](https://www.codacy.com/app/triplez/open-event-orga-server/file/2406546261/issues/source?bid=3233864&fileBranchId=3233880#l22)

File: open_event/models/reviews.py [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    def serialize(self):
+        """Return object data in easily serializeable format"""
+        return {'id': self.id,
+                'email': self.email,
+                'comment': self.comment,
+                'rating': self.rating,
+                'session_id': self.session_id}
+
+    def __repr__(self):
+        return '<Review %r>' % (self.id)
```
Comment (juslee) [inline open_event/models/reviews.py]: Codacy Issue found: [Final newline missing](https://www.codacy.com/app/triplez/open-event-orga-server/file/2410978632/issues/source?bid=3233864&fileBranchId=3233880#l32)

File: open_event/models/reviews.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
+from . import db
+
+class Review(db.Model):
```
Comment (juslee) [inline open_event/models/reviews.py]: Codacy Issue found: [Too few public methods (1/2)](https://www.codacy.com/app/triplez/open-event-orga-server/file/2410978632/issues/source?bid=3233864&fileBranchId=3233880#l3)

File: open_event/models/reviews.py [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    """Review model class"""
+    __tablename__ = 'session_reviews'
+    id = db.Column(db.Integer, primary_key=True)
```
Comment (juslee) [inline open_event/models/reviews.py]: Codacy Issue found: [Invalid class attribute name "id"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2410978632/issues/source?bid=3233864&fileBranchId=3233880#l6)

File: open_event/helpers/data.py [hunk 14]
```diff
@@ -71,11 +72,47 @@ def remove_track(track_id):
         Track will be removed from database
         :param track_id: Track id to remove object
         """
+        Session.query.filter_by(track_id=track_id).delete()
         track = Track.query.get(track_id)
         delete_from_db(track, "Track deleted")
         flash('You successfully deleted track')
 
     @staticmethod
+    def create_review(data, session_id, is_accepted=True):
```
Comment (juslee) [inline open_event/helpers/data.py]: Codacy Issue found: [Unused argument 'is_accepted'](https://www.codacy.com/app/triplez/open-event-orga-server/file/2410978618/issues/source?bid=3233864&fileBranchId=3233880#l81)

File: migrations/versions/0b5368c41418_.py [hunk 15]
```diff
@@ -0,0 +1,34 @@
+"""empty message
```
Comment (juslee) [inline migrations/versions/0b5368c41418_.py]: Codacy Issue found: [Invalid module name "0b5368c41418_"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2410978595/issues/source?bid=3233864&fileBranchId=3233880#l1)

File: migrations/versions/0b5368c41418_.py [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+Revision ID: 0b5368c41418
+Revises: 202b15987640
+Create Date: 2016-04-09 16:25:09.139313
+
+"""
+
+# revision identifiers, used by Alembic.
+revision = '0b5368c41418'
```
Comment (juslee) [inline migrations/versions/0b5368c41418_.py]: Codacy Issue found: [Invalid constant name "revision"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2410978595/issues/source?bid=3233864&fileBranchId=3233880#l10)

File: migrations/versions/0b5368c41418_.py [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+down_revision = '202b15987640'
```
Comment (juslee) [inline migrations/versions/0b5368c41418_.py]: Codacy Issue found: [Invalid constant name "down_revision"](https://www.codacy.com/app/triplez/open-event-orga-server/file/2410978595/issues/source?bid=3233864&fileBranchId=3233880#l11)

**Comment by SaptakS** [comment]
I am not sure why the Travis CI build fails.

**Comment by juslee** [comment]
Because the travis.yml hasn't been created yet. I need someone to help me set this up.

**Comment by mananwason** [comment]
I'll do it @juslee :)

**Pull request closed without merging**